CP4AIOPS-3113 Introduce configurable delimiter for LDAP group names

kbrock opened the pull request:
## Overview

|Auth Server|Interface|Delimiter|
|-----------|---------|---------|
| Liberty   |     oicd|      `,`|
|           |     saml|      `;`|
|ipa        |     sssd|      `:`|
|ldap       |     sssd|      `:`|

## Goal

Allow users to have a `:` in group names.

Since the group can come in with 3 possible delimiters, we treat `:`, `,`, and `;` as possible delimiters.

This works fine if groups do not have one of those characters in it.

Liberty uses a `,` as a delimiter and allows `:` in the group name.
If a liberty server sends groups value of `'group1,people:fun,people:sad'`, we treat this as groups `'group1'`, `'people'`, `'fun'`, `'people'`, and `'sad'`.

If a configuration references `group1`, then it works fine, but configuration that references `people:fun` will not find the group and not work as expected.

## Solution

1. Allow apache to pass the delimiter used to ruby.
2. Allow the admin to override this delimiter in advanced settings.

The various configurations use different delimiters, so it makes sense to allow each of those configurations to specify the delimiter that is being used.

If an oidc server uses a different delimiter, then allow the admin to configure the correct delimiter in settings.

## See Also

blocked:

- [x] https://github.com/ManageIQ/manageiq-appliance_console/pull/265
- [x] https://github.com/ManageIQ/manageiq-appliance/pull/389
- [x] Release of appliance_console gem version 10.0.0 

dependent:

- [x] https://github.com/ManageIQ/manageiq-pods/pull/1155

ref:

- https://github.com/ManageIQ/manageiq/pull/32
- update: https://github.com/ManageIQ/httpd_configmap_generator


## Next steps

We could change the configuration for `LookupUserGroup` and `mod_auth_mellon` to just use a `,` as a delimiter, but that felt like too much change.

bdunne commented on app/models/user.rb:
```diff
@@ -225,11 +225,11 @@ def miq_user_role_name
   end
 
-  def self.authenticator(username = nil)
+  def self.authenticator(username = nil) ##
```
Were these supposed to be here?

Fryguy commented on app/models/authenticator/httpd.rb:
```diff
@@ -146,7 +150,7 @@ def user_details_from_headers(username, request)
       end
 
-      groups     = CGI.unescape(user_headers['X-REMOTE-USER-GROUPS'] || '').split(/[;:,]/)
+      groups     = CGI.unescape(user_headers['X-REMOTE-USER-GROUPS'] || '').split(/[#{self.class.group_delimiters}]/)
```
Since this can come from user input, I think we need to be more careful here.  Thinking:

```suggestion
      groups     = CGI.unescape(user_headers['X-REMOTE-USER-GROUPS'] || '').split(/[#{Regexp.escape(self.class.group_delimiters)}]/)
```

Fryguy replied:
For example, what happens if a user puts `]` (because they can)

jrafanie replied:
yeah, I'm not sure how to do it.  We probably need to worry about typos and security issues with the input.

kbrock replied:
solution: treat the delimiter as static text.

1. We will allow multiple characters
2. We will not need special escaping

Fryguy commented on config/settings.yml:
```diff
@@ -16,6 +16,8 @@
   :ldapport: '389'
   :mode: database
+  :http_group_delimiters: ";:,"
+
```
bonus newline here should be moreved.

kbrock replied:
done.
also, another commit actually blanked this value out.

Fryguy commented on app/models/authenticator/httpd.rb:
```diff
@@ -4,6 +4,10 @@ def self.proper_name
     end
 
+    def self.group_delimiters
+      ::Settings.authentication.http_group_delimiters
```
http doesn't make sense, but maybe `httpd` does or perhaps just remove the prefix altogether.

```suggestion
      ::Settings.authentication.httpd_group_delimiters
```

or

```suggestion
      ::Settings.authentication.group_delimiters
```

jrafanie replied:
haha, maybe `external_group_delimiters`... naming is hard

Fryguy commented on app/models/authenticator/httpd.rb:
```diff
@@ -146,7 +150,7 @@ def user_details_from_headers(username, request)
       end
 
-      groups     = CGI.unescape(user_headers['X-REMOTE-USER-GROUPS'] || '').split(/[;:,]/)
+      groups     = CGI.unescape(user_headers['X-REMOTE-USER-GROUPS'] || '').split(/[#{self.class.group_delimiters}]/)
```
Actually, now that I wrote that we have to be resilient to users breaking stuff.  What if they set it to blank.

Also do we need to handle consecutive delimiters?  I had to handle all these cases in [floe's StringSplit](https://github.com/ManageIQ/floe/blob/master/lib/floe/workflow/intrinsic_function/transformer.rb#L235-L242), hence why I'm asking.

kbrock replied:
I am not able to track down how to configure ldap group separators for Liberty.
The only delimiters I have been able to find in the Liberty xml configuration are single character.

Fryguy replied:
KeyCloak is our other reference system - can you set these values in KeyCloak?

kbrock replied:
I see a bunch of double letter delimiters in keycloak.
Diving deeper

kbrock replied:
ok. if the user sets to blank (the default value), then we will allow any of the 3 (the current behavior)

jrafanie commented on config/settings.yml:
```diff
@@ -16,6 +16,7 @@
   :ldaphost:
   :ldapport: '389'
   :mode: database
+  :group_delimiters: ";:,"
```
@kbrock would it simplify things to expose only a single character as a group delimiter per authenticator type, or perhaps for all authenticators?

Would that decrease the opportunity for a regex security or a typo?

In other words, something like this but please clean it up:

```ruby
groups = if httpd_group_delimiter.to_s.length == 1
  CGI.unescape(user_headers['X-REMOTE-USER-GROUPS'] || '').split(httpd_group_delimiter.to_s)
else
  CGI.unescape(user_headers['X-REMOTE-USER-GROUPS'] || '').split(/[;:,]/)
end

jrafanie replied:
Note, I thought it was weird that we support multiple delimiters based on whichever ldap system we were adding at the time.  It feels like each "type" could have the current values.  I'm not sure which uses `;` or `:` or `,`.  Either way, if we keep the existing behavior as the default if you don't specify a delimiter, it shouldn't break anything unless you opt-in and use the wrong value for your system.

kbrock replied:
updated comment to describe which system uses which delimiter
Also modified the configuration to make it more obvious

kbrock commented:
update:
- rebased
- reindented hash rockets in specs (rubocop)

kbrock commented:
Updated:
- Rebased
- No longer defaulting the delimiter settings.yml value.
- Apache now sends a delimiter value for all configurations.
- The settings takes precedence over the apache header.

kbrock commented:
update:
- Require appliance_console 9.2 (it can handle the 2 different saml configuration files now stored in appliance
- group parameter now called `REMOTE_USER_GROUP_DELIMITER`
- setting is now called `group_delimiter` (it was called delimiters, but it is just a single delimiter now)

Fryguy commented on spec/models/authenticator/httpd_spec.rb:
```diff
@@ -707,17 +719,55 @@ def authenticate
+          super().merge('X-Remote-User-Groups' => 'wibble:wobble@fqdn,bubble@fqdn')
+        end
+
+        it "handles a group with a colon" do
```
```suggestion
        it "handles a group with a comma" do
```

Also the other `it`s below have the wrong term as well.

Fryguy commented on spec/models/authenticator/httpd_spec.rb:
```diff
@@ -707,17 +719,55 @@ def authenticate
+        end
+      end
+
+      context "using custom delimiter in settings" do
```
This context name is identical to the other context.

Fryguy commented:
I noticed you have a test that checks priority of config over header, however I don't see a test that shows the header working correctly without config.  There are a few other tests I think needed.

- a test that shows a header working without config set
- a test where the header is _not_ set showing the fallback to the regex
- update existing tests to have the header passed in even if there are no groups (because the env is blank in the .conf files)

Fryguy commented on spec/models/authenticator/httpd_spec.rb:
```diff
@@ -707,17 +719,55 @@ def authenticate
+      context "using header custom delimiter (prioritize config over header)" do
+        let(:config) { {:httpd_role => true, :group_delimiters => "|"} }
+        let(:headers) do
+          super().merge('X-Remote-User-Groups' => 'wibble:wobble@fqdn|bubble@fqdn', 'X-REMOTE-USER-GROUP-DELIMITER' => ",")
```
Minor, but the header case should match the other one, so this should probably be

```suggestion
          super().merge('X-Remote-User-Groups' => 'wibble:wobble@fqdn|bubble@fqdn', 'X-Remote-User-Group-Delimiter' => ",")
```

kbrock commented:
update:
- removed duplicate custom delimiter test (, vs |)
- changed spec wording so it was clear whether we were using setting, fallback, or header.
- added spec to test setting override without header.
- changed case of the header
- fixed specs to call setting `group_delimiter` (was plural)

Fryguy commented:
Backported to `radjabov` in commit 8859562fce4177f65287669bbc61c1cb22800d04.

```text
commit 8859562fce4177f65287669bbc61c1cb22800d04
Author: Jason Frey <fryguy9@gmail.com>
Date:   Tue Nov 19 15:41:30 2024 -0500

    Merge pull request #23139 from kbrock/CP4AIOPS-3113
    
    CP4AIOPS-3113 Introduce configurable delimiter for LDAP group names
    
    (cherry picked from commit e88fdc670a6c34788becabe1c096680d04c549b9)
```
